src/link.js

```diff
@@ -0,0 +1,259 @@
+ */
+
+import Feature from '../core/feature.js';
+import ClickObserver from '../engine/view/observer/clickobserver.js';
+import LinkEngine from './linkengine.js';
+import LinkElement from './linkelement.js';
+
+import Model from '../ui/model.js';
+
+import ButtonController from '../ui/button/button.js';
+import ButtonView from '../ui/button/buttonview.js';
+
+import LinkBalloonPanel from './ui/linkballoonpanel.js';
+import LinkBalloonPanelView from './ui/linkballoonpanelview.js';
+
+/**
+ * The link feature.
+ *
+ * It uses the {@link basic-styles.LinkEngine link engine feature}.
+ *
+ * @memberOf link
+ * @extends core.Feature
+ */
+export default class Link extends Feature {
+	/**
+	 * @inheritDoc
+	 */
+	static get requires() {
+		return [ LinkEngine ];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const viewDocument = this.editor.editing.view;
+
+		// Register document click observer.
+		viewDocument.addObserver( ClickObserver );
+
+		/**
+		 * Link balloon panel component.
+		 *
+		 * @member {link.ui.LinkBalloonPanel}
+		 */
+		this.balloonPanel = this._createBalloonPanel();
+
+		// Create toolbar buttons.
+		this._createToolbarLinkButton();
+		this._createToolbarUnlinkButton();
+	}
+
+	/**
+	 * Create toolbar link button. Click on button will show
+	 * {@link link.ui.LinkBalloonPanel LinkBalloonPanel} attached to the selection.
+	 *
+	 * @private
+	 */
+	_createToolbarLinkButton() {
+		const editor = this.editor;
+		const viewDocument = editor.editing.view;
+		const linkCommand = editor.commands.get( 'link' );
+		const t = editor.t;
+
+		// Create button model.
+		const linkButtonModel = new Model( {
+			isEnabled: true,
+			isOn: false,
+			label: t( 'Link' ),
+			icon: 'link',
+			keystroke: 'CTRL+l'
+		} );
+
+		// Bind button model to command.
+		linkButtonModel.bind( 'isEnabled' ).to( linkCommand, 'isEnabled' );
+
+		// Show Balloon Panel on button click only when editor is focused.
+		this.listenTo( linkButtonModel, 'execute', () => {
+			if ( !viewDocument.isFocused ) {
+				return;
+			}
+
+			this._attachPanelToElement();
+		} );
+
+		// Add link button to feature components.
+		editor.ui.featureComponents.add( 'link', ButtonController, ButtonView, linkButtonModel );
+	}
+
+	/**
+	 * Create toolbar unlink button. Click on button will unlink selected link.
+	 *
+	 * @private
+	 */
+	_createToolbarUnlinkButton() {
+		const editor = this.editor;
+		const t = editor.t;
+		const unlinkCommand = editor.commands.get( 'unlink' );
+
+		// Create button model.
+		const unlinkButtonModel = new Model( {
+			isEnabled: false,
+			isOn: false,
+			label: t( 'Unlink' ),
+			icon: 'unlink'
+		} );
+
+		// Bind button model to command.
+		unlinkButtonModel.bind( 'isEnabled' ).to( unlinkCommand, 'hasValue' );
+
+		// Execute unlink command and hide panel if is opened.
+		this.listenTo( unlinkButtonModel, 'execute', () => {
+			editor.execute( 'unlink' );
+
+			if ( this.balloonPanel && this.balloonPanel.view.isVisible ) {
```

Comment:
Why checking `this.balloonPanel &&`? Is it possible that the panel does not exist at this point?